File: test/php/widget/ApplyOttTest.php
<?php

namespace DanaUat\Widget;

use PHPUnit\Framework\TestCase;
use Dana\Widget\v1\Api\WidgetApi;
use Dana\Configuration;
use Dana\ObjectSerializer;
use Dana\Env;
use Dana\ApiException;
use DanaUat\Helper\Assertion;
use DanaUat\Helper\Util;
use DanaUat\Widget\OauthUtil;
use Exception;
use GuzzleHttp\Psr7\Response;

class ApplyOttTest extends TestCase
{
    private static $titleCase = 'ApplyOtt';
    private static $jsonPathFile = 'resource/request/components/Widget.json';
    private static $apiInstance;
    private static $phoneNumber = '0811742234';
    private static $userPin = '123321';
    private static $authCode;
    private static $accessToken;

    public static function setUpBeforeClass(): void
    {
        $configuration = new Configuration();
        $configuration->setApiKey('PRIVATE_KEY', getenv('PRIVATE_KEY'));
        $configuration->setApiKey('ORIGIN', getenv('ORIGIN'));
        $configuration->setApiKey('X_PARTNER_ID', getenv('X_PARTNER_ID'));
        $configuration->setApiKey('ENV', Env::SANDBOX);
        self::$apiInstance = new WidgetApi(null, $configuration);
        
        // Get OAuth authorization code
        self::$authCode = OauthUtil::getAuthCode(
            getenv('X_PARTNER_ID'),
            null,
            self::$phoneNumber,
            self::$userPin,
            null
        );
        
        echo "\nObtained auth code: " . self::$authCode . "\n";
        
        // Get access token using the auth code
        $tokenCaseName = 'ApplyTokenSuccess';
        $tokenJsonDict = Util::getRequest(
            'resource/request/components/Widget.json',
            'ApplyToken',
            $tokenCaseName
        );
        
        $tokenRequestObj = ObjectSerializer::deserialize(
            $tokenJsonDict,
            'Dana\Widget\v1\Model\ApplyTokenRequest'
        );
        
        $tokenRequestObj->setAuthCode(self::$authCode);
        
        try {
            $apiResponse = self::$apiInstance->applyToken($tokenRequestObj);
            $responseJson = json_decode($apiResponse->__toString(), true);
            self::$accessToken = $responseJson['accessToken'];
            echo "\nObtained access token: " . self::$accessToken . "\n";
        } catch (Exception $e) {
            echo "\nFailed to obtain access token: " . $e->getMessage() . "\n";
        }
    }

    /**
     * Should give success response for apply ott scenario
     */
    public function testApplyOttSuccess(): void
    {

        Util::withDelay(function () {
            $caseName = 'ApplyOttSuccess';
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );
            $jsonDict['additionalInfo']['accessToken'] = self::$accessToken;
            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\ApplyOTTRequest'
            );

            $apiResponse = self::$apiInstance->applyOTT($requestObj);
            
            Assertion::assertResponse(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName,
                $apiResponse->__toString()
            );
        });
    }

    /**
     * Should fail to apply OTT with token not found
     */
    public function testApplyOttFailTokenNotFound(): void
    {

        Util::withDelay(function () {
            $caseName = 'ApplyOttCustomerTokenNotFound';
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );
            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\ApplyOTTRequest'
            );
            // Set an invalid access token for testing token not found error
            $jsonDict['additionalInfo']['accessToken'] = 'invalid_access_token_for_testing';
            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\ApplyOTTRequest'
            );
            try {
                self::$apiInstance->applyOTT($requestObj);
                $this->fail('Expected ApiException was not thrown');
            } catch (ApiException $e) {
                Assertion::assertFailResponse(self::$jsonPathFile, self::$titleCase, $caseName, $e->getResponseBody());
                $this->assertTrue(true);
            }
        });
    }

    /**
     * Should fail to apply OTT with invalid user status
     */
    public function testApplyOttFailInvalidUserStatus(): void
    {

        $this->markTestSkipped('Scenario skipped because the result the same with testApplyOttFailTokenNotFound');
        
        Util::withDelay(function () {
            $caseName = 'ApplyOttFailInvalidUserStatus';
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );
            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\ApplyOTTRequest'
            );
            // Set a modified access token that will trigger invalid user status error
            $jsonDict['additionalInfo']['accessToken'] = self::$accessToken !== null ? self::$accessToken.'_modified' : 'test_access_token';
            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\ApplyOTTRequest'
            );
            try {
                self::$apiInstance->applyOTT($requestObj);
                $this->fail('Expected ApiException was not thrown');
            } catch (ApiException $e) {
                Assertion::assertFailResponse(self::$jsonPathFile, self::$titleCase, $caseName, $e->getResponseBody());
                $this->assertTrue(true);
            }
        });
    }
}

File: test/python/widget/apply_ott_test.py
import os
import pytest
import asyncio
from dana.utils.snap_configuration import SnapConfiguration, AuthSettings, Env
from dana.widget.v1.enum import *
from dana.widget.v1.models import *
from dana.widget.v1 import *
from dana.widget.v1.api import *
from dana.api_client import ApiClient
from dana.exceptions import *
from uuid import uuid4
from helper.util import get_request, with_delay
from helper.assertion import assert_response, assert_fail_response
from widget.automate_oauth import automate_oauth

title_case = "ApplyOtt"
json_path_file = "resource/request/components/Widget.json"

configuration = SnapConfiguration(
    api_key=AuthSettings(
        PRIVATE_KEY=os.environ.get("PRIVATE_KEY"),
        ORIGIN=os.environ.get("ORIGIN"),
        X_PARTNER_ID=os.environ.get("X_PARTNER_ID"),
        ENV=Env.SANDBOX,
        DANA_ENV=Env.SANDBOX,
        X_DEBUG="true"
    )
)

with ApiClient(configuration) as api_client:
    api_instance = WidgetApi(api_client)

def generate_partner_reference_no():
    return str(uuid4())

def get_access_token(phone_number=None, pin=None):
    if os.environ.get("CI") == "true":
        pytest.skip("Skipped in CI/CD")
    auth_code = asyncio.run(automate_oauth(phone_number=phone_number, pin=pin))
    print(auth_code)
    access_token = apply_token(auth_code)
    return access_token

@pytest.fixture(scope="module")
def test_access_token_success():
    return get_access_token(phone_number="083811223355", pin="181818")

@pytest.fixture(scope="module")
def test_access_token_user_abnormal():
    return get_access_token(phone_number="0855100800", pin="146838")

@with_delay()
def test_apply_ott_success(test_access_token_success):
    if os.environ.get("CI") == "true":
        pytest.skip("Skipped in CI/CD")
    case_name = "ApplyOttSuccess"
    access_token = test_access_token_success
    json_dict = get_request(json_path_file, title_case, case_name)
    applyOTTRequestAdditionalInfo = ApplyOTTRequestAdditionalInfo.from_dict(json_dict.get("additionalInfo", {}))
    applyOTTRequestAdditionalInfo.access_token = access_token
    applyOTTRequestAdditionalInfo.device_id = "deviceid123"
    json_dict["additionalInfo"] = applyOTTRequestAdditionalInfo.to_dict()
    apply_ott_request_obj = ApplyOTTRequest.from_dict(json_dict)
    api_response = api_instance.apply_ott(apply_ott_request_obj)
    assert_response(json_path_file, title_case, case_name, ApplyOTTResponse.to_json(api_response))

@with_delay()
def test_apply_ott_fail_token_not_found():
    case_name = "ApplyOttFailTokenNotFound"
    json_dict = get_request(json_path_file, title_case, case_name)
    applyOTTRequestAdditionalInfo = ApplyOTTRequestAdditionalInfo.from_dict(json_dict.get("additionalInfo", {}))
    applyOTTRequestAdditionalInfo.access_token = "GtRLpA0TyqK3becMq4dCMnVf1N9KLHNixVfC1800"
    applyOTTRequestAdditionalInfo.device_id = "deviceid123"
    json_dict["additionalInfo"] = applyOTTRequestAdditionalInfo.to_dict()
    apply_ott_request_obj = ApplyOTTRequest.from_dict(json_dict)

    try:
        api_response = api_instance.apply_ott(apply_ott_request_obj)
        assert_response(json_path_file, title_case, case_name, ApplyOTTResponse.to_json(api_response))
    except Exception as e:
        print("Error occurred while processing request:", e)
        
@with_delay()
def test_apply_ott_fail_token_user_abnormal(test_access_token_user_abnormal):
    if os.environ.get("CI") == "true":
        pytest.skip("Skipped in CI/CD")
    case_name = "ApplyOttCustomerAccountUserStatusAbnormal"
    json_dict = get_request(json_path_file, title_case, case_name)
    applyOTTRequestAdditionalInfo = ApplyOTTRequestAdditionalInfo.from_dict(json_dict.get("additionalInfo", {}))
    applyOTTRequestAdditionalInfo.access_token = test_access_token_user_abnormal
    applyOTTRequestAdditionalInfo.device_id = "deviceid123"
    json_dict["additionalInfo"] = applyOTTRequestAdditionalInfo.to_dict()
    apply_ott_request_obj = ApplyOTTRequest.from_dict(json_dict)

    try:
        api_response = api_instance.apply_ott(apply_ott_request_obj)
        assert_response(json_path_file, title_case, case_name, ApplyOTTResponse.to_json(api_response))
    except ForbiddenException as e:
        print("Error occurred while processing request:", e)

def apply_token(auth_code):
    json_dict = get_request(json_path_file, "ApplyToken", "ApplyTokenSuccess")
    # Use the provided authorization code from the fixture
    json_dict["authCode"] = auth_code

    # Create the request object from the JSON dictionary
    request_obj = ApplyTokenAuthorizationCodeRequest.from_dict(json_dict)
    response = api_instance.apply_token(request_obj)
    return response.accessToken